votor/src/event_handler/stats.rs

```diff
@@ -0,0 +1,312 @@
+            } else if vote_type == VoteType::Skip {
+                let entry = self.slot_tracking_map.entry(vote.vote.slot()).or_default();
+                entry.vote_skip = Some(Instant::now());
+            }
+        } else {
+            warn!("Unexpected BLS operation: {bls_op:?}");
+        }
+    }
+
+    pub fn maybe_report(&mut self) {
+        let now = Instant::now();
+        if now.duration_since(self.last_report_time) < STATS_REPORT_INTERVAL {
+            return;
+        }
+        datapoint_info!(
+            "event_handler_stats",
+            ("ignored", self.ignored as i64, i64),
+            (
+                "leader_window_replaced",
+                self.leader_window_replaced as i64,
+                i64
+            ),
+            ("set_root_count", self.set_root_count as i64, i64),
+            ("timeout_set", self.timeout_set as i64, i64),
+        );
+        for (event, EventCountAndTime { count, time_us }) in &self.received_events_count_and_timing
+        {
+            datapoint_info!(
+                "event_handler_received_event_count_and_timing",
+                ("event", format!("{:?}", event), String),
+                ("count", *count as i64, i64),
+                ("elapsed_us", *time_us as i64, i64)
+            );
+        }
+        datapoint_info!(
+            "event_handler_timing",
+            (
+                "receive_event_time_us",
+                self.receive_event_time_us as i64,
+                i64
+            ),
+            (
+                "send_votes_batch_time_us",
+                self.send_votes_batch_time_us as i64,
+                i64
+            ),
+        );
+        for (vote_type, count) in &self.sent_votes {
+            datapoint_info!(
+                "event_handler_sent_vote_count",
+                ("vote", format!("{:?}", vote_type), String),
+                ("count", *count as i64, i64)
+            );
+        }
+        // Only report if the slot is lower than root_slot
+        let split_off_map = self.slot_tracking_map.split_off(&self.root_slot);
+        for (slot, tracking) in &self.slot_tracking_map {
+            let start = tracking.start;
+            datapoint_info!(
+                "event_handler_slot_tracking",
+                ("slot", *slot as i64, i64),
+                (
+                    "first_shred",
+                    tracking.first_shred.map(|t| {
+                        t.saturating_duration_since(start)
+                            .as_micros()
+                            .min(i64::MAX as u128) as i64
+                    }),
+                    Option<i64>
+                ),
+                (
+                    "parent_ready",
+                    tracking.parent_ready.map(|t| {
+                        t.saturating_duration_since(start)
+                            .as_micros()
+                            .min(i64::MAX as u128) as i64
+                    }),
```

Comment:
nit: If we repeat this one more time I'd say put it in a helper

Reply:
built out an `impl SlotTracking` to abstract a lot of this away